staking/slash/double-sign.go

```diff
@@ -145,66 +145,70 @@ var (
 // applySlashRate returns (amountPostSlash, amountOfReduction, amountOfReduction / 2)
 func applySlashRate(
 	amount *big.Int, rate numeric.Dec,
-) (*big.Int, *big.Int, *big.Int) {
+) (*big.Int, *big.Int) {
 	amountPostSlash := numeric.NewDecFromBigInt(
 		amount,
 	).Mul(numeric.OneDec().Sub(rate)).TruncateInt()
 	amountOfSlash := new(big.Int).Sub(amount, amountPostSlash)
-	return amountPostSlash, amountOfSlash, new(big.Int).Div(amountOfSlash, common.Big2)
+	return amountOfSlash, new(big.Int).Div(amountOfSlash, common.Big2)
 }
 
 func delegatorSlashApply(
-	delegatedToValidator *staking.ValidatorWrapper,
+	snapshot, current *staking.ValidatorWrapper,
 	rate numeric.Dec,
 	state *state.DB,
 	reporter common.Address,
 	doubleSignEpoch *big.Int,
 	slashTrack *Application,
 ) error {
-	for _, delegationSnapshot := range delegatedToValidator.Delegations {
-		postSlashBal, slashDebt, halfOfSlashDebt := applySlashRate(
+	for _, delegationSnapshot := range snapshot.Delegations {
+		slashDebt, halfOfSlashDebt := applySlashRate(
 			delegationSnapshot.Amount, rate,
 		)
-		if postSlashBal.Sign() == -1 {
-			stillOwe := new(big.Int).Sub(slashDebt, delegationSnapshot.Amount)
-			if wrapper := state.GetStakingInfo(
-				delegatedToValidator.Address,
-			); wrapper != nil {
-				for _, delegationNow := range wrapper.Delegations {
-					if delegationNow.Hash() == delegationSnapshot.Hash() {
-						for _, undelegate := range delegationNow.Undelegations {
-							// the epoch matters, only those undelegation
-							// such that epoch>= doubleSignEpoch should be slashable
-							if undelegate.Epoch.Cmp(doubleSignEpoch) >= 0 {
-								continue
-							}
-							if stillOwe.Cmp(common.Big0) <= 0 {
-								// paid off the slash debt
-								break
-							}
-							amtPostSlash, amtSlash, halfOfSlashAmt := applySlashRate(
-								undelegate.Amount, rate,
-							)
-							undelegate.Amount = amtPostSlash
-							stillOwe.Sub(stillOwe, halfOfSlashAmt)
-							state.AddBalance(reporter, halfOfSlashAmt)
-							slashTrack.TotalSlashed.Add(slashTrack.TotalSlashed, amtSlash)
-							slashTrack.TotalSnitchReward.Add(slashTrack.TotalSnitchReward, halfOfSlashAmt)
+		h := delegationSnapshot.Hash()
+		for _, delegationNow := range current.Delegations {
+			if nowAmt := delegationNow.Amount; delegationNow.Hash() == h {
+				state.AddBalance(reporter, halfOfSlashDebt)
+				slashTrack.TotalSnitchReward.Add(
+					slashTrack.TotalSnitchReward, halfOfSlashDebt,
+				)
+				const (
+					haveEnoughToPayOff               = 1
+					paidOffExact                     = 0
+					debtCollectionsRepoUndelegations = -1
+				)
+				switch d := new(big.Int).Sub(nowAmt, halfOfSlashDebt); d.Sign() {
```

Comment:
Should compare to slashDebt. halfOfSlashDebt is only used to credit reporter. Other than that, it's all about whole slashDebt.

Reply:
I see, so credit reporter is getting token out of thin air, the wording on slashing doc was a little ambiguous to me.